File: engine/src/ingest/prices.ts
/** Price ingestion — port of ingest/prices.py.
 *
 * Python leans on `yfinance`; here we call the endpoint yfinance itself wraps
 * (Yahoo's chart API) and reproduce the same derivation: last close, previous
 * close for the 1-day change, and the quote currency. Verified against
 * yfinance output — identical closes, price and change for the same symbol.
 *
 * On failure we fall back to the last stored price and mark it stale, so the
 * digest can say so honestly rather than silently using old numbers.
 */

import type { PriceSnapshot } from "../models.ts";
import type { HttpClient } from "../platform.ts";
import { pyRound } from "../round.ts";
import type { Store } from "../store/db.ts";

const CHART = "https://query1.finance.yahoo.com/v8/finance/chart/";

interface ChartResponse {
  chart?: {
    result?: Array<{
      meta?: { currency?: string; symbol?: string };
      indicators?: { quote?: Array<{ close?: Array<number | null> }> };
    }>;
  };
}

/** Parse a Yahoo chart payload the same way ingest/prices.py derives its values. */
export function parseChart(ticker: string, json: unknown): PriceSnapshot | null {
  const res = (json as ChartResponse)?.chart?.result?.[0];
  if (!res) return null;
  const raw = res.indicators?.quote?.[0]?.close ?? [];
  const closes = raw.filter((v): v is number => v !== null && v !== undefined && Number.isFinite(v));
  if (closes.length === 0) return null;
  const price = closes[closes.length - 1];
  let change: number | null = null;
  if (closes.length >= 2) {
    const prev = closes[closes.length - 2];
    if (prev) change = pyRound(((price - prev) / prev) * 100, 2);
  }
  return {
    ticker: ticker.toUpperCase(),
    price: pyRound(price, 4),
    currency: res.meta?.currency ?? "",
    change_pct_1d: change,
    stale: false,
  };
}

async function fetchOne(http: HttpClient, ticker: string): Promise<PriceSnapshot | null> {
  const url = `${CHART}${encodeURIComponent(ticker)}?range=5d&interval=1d`;
  const json = await http.getJson(url);
  if (!json) return null;
  return parseChart(ticker, json);
}

/** Live fetch with per-ticker fallback to the cached price (marked stale). */
export async function fetchPrices(
  tickers: string[], http: HttpClient, store: Store,
): Promise<Record<string, PriceSnapshot>> {
  const out: Record<string, PriceSnapshot> = {};
  for (const ticker of tickers) {
    const snap = await fetchOne(http, ticker);
    if (snap === null) {
      const cached = await cachedPrice(ticker, store);
      if (cached) out[ticker.toUpperCase()] = cached;
      continue;
    }
    await store.savePrice({
      ticker: snap.ticker, price: snap.price, currency: snap.currency,
      change_pct_1d: snap.change_pct_1d, as_of: new Date().toISOString(),
    });
    out[ticker.toUpperCase()] = snap;
  }
  return out;
}

async function cachedPrice(ticker: string, store: Store): Promise<PriceSnapshot | null> {
  const row = await store.latestPrice(ticker);
  if (!row) return null;
  return {
    ticker: row.ticker, price: row.price, currency: row.currency ?? "",
    change_pct_1d: row.change_pct_1d, stale: true, // came from cache
  };
}

/** Last-known prices only, no network — the fast path for the UI. */
export async function cachedPrices(
  tickers: string[], store: Store,
): Promise<Record<string, PriceSnapshot>> {
  const out: Record<string, PriceSnapshot> = {};
  for (const t of tickers) {
    const c = await cachedPrice(t, store);
    if (c) out[t.toUpperCase()] = c;
  }
  return out;
}

File: src/investraton/ingest/prices.py
"""Free price data via yfinance, with graceful degradation to cached values.

If a fetch fails (network down, rate limited, bad ticker) we fall back to the
last stored price and mark it stale, so the digest can say so honestly rather
than silently using old numbers.
"""

from __future__ import annotations

import logging

from ..models import PriceSnapshot, utcnow
from ..store import Store

log = logging.getLogger(__name__)


def fetch_prices(tickers: list[str], store: Store) -> dict[str, PriceSnapshot]:
    out: dict[str, PriceSnapshot] = {}
    try:
        import yfinance as yf
    except ImportError:
        log.warning("yfinance not installed; using cached prices only.")
        return _all_cached(tickers, store)

    for ticker in tickers:
        snap = _fetch_one(yf, ticker)
        if snap is None:
            cached = store.latest_price(ticker)
            if cached is not None:
                out[ticker.upper()] = cached
            continue
        store.save_price(snap)
        out[ticker.upper()] = snap
    return out


def _fetch_one(yf, ticker: str) -> PriceSnapshot | None:
    try:
        t = yf.Ticker(ticker)
        hist = t.history(period="5d")
        if hist is None or hist.empty:
            return None
        closes = hist["Close"].dropna()
        if closes.empty:
            return None
        price = float(closes.iloc[-1])
        change = None
        if len(closes) >= 2:
            prev = float(closes.iloc[-2])
            if prev:
                change = round((price - prev) / prev * 100.0, 2)
        currency = ""
        try:
            currency = (t.fast_info.get("currency") or "") if hasattr(t, "fast_info") else ""
        except Exception:  # noqa: BLE001
            currency = ""
        return PriceSnapshot(
            ticker=ticker.upper(),
            price=round(price, 4),
            currency=currency,
            change_pct_1d=change,
            as_of=utcnow(),
            stale=False,
        )
    except Exception as e:  # noqa: BLE001 - external
        log.warning("price fetch failed for %s: %s", ticker, e)
        return None


def cached_prices(tickers: list[str], store: Store) -> dict[str, PriceSnapshot]:
    """Read last-known prices from the store only (no network). Fast path for the UI."""
    return _all_cached(tickers, store)


def _all_cached(tickers: list[str], store: Store) -> dict[str, PriceSnapshot]:
    out: dict[str, PriceSnapshot] = {}
    for ticker in tickers:
        cached = store.latest_price(ticker)
        if cached is not None:
            out[ticker.upper()] = cached
    return out
